### ISSUE TITLE
zero sized arrays not correctly serialized?

### ISSUE THREAD

**Issue by wds15** [stan-dev/rstan#371]
#### Summary:
stan_rdump serialized arrays with zero size  cannot be read with cmdstan

#### Description:
Serializing with stan_rdump and starting a cmdstan run does not work any more if the arrays has size zero. This used to work with rstan 2.12 and cmdstand 2.12.

#### Reproducible Steps:
dummy data:
```
library(rstan)

N <- 10
y <- array(0, 10)
stan_rdump(list("N", "y"), "ok.R")

N <- 0
y <- array(NA, 0)
stan_rdump(list("N", "y"), "notok.R")
```

sample stan program
```
data {
  int N;
  int y[N];
}
parameters {
  real mu;
}
model {
  target += normal_lpdf(mu|0, 1);
}
```

#### Current Output:
If applicable, any relevant output from RStan.

#### Expected Output:
Error from cmdstan:
```
[weberse2@node04:~/work/stan/misc] ./stan_int_bug sample data file=notok.R

Exception: data y value  beyond int range
Diagnostic information: 
stan_2.13.1/src/stan/io/dump.hpp(518): Throw in function bool stan::io::dump_reader::next()
Dynamic exception type: boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<std::invalid_argument> >
std::exception::what: data y value  beyond int range
```

#### RStan Version:
2.13.2

#### R Version:
3.2.3

#### Operating System:
RHEL 6.7

**Comment by maverickg** [comment]
see https://github.com/stan-dev/rstan/issues/366

**Comment by wds15** [comment]
ah... so does this mean that the problem is with cmdstan? I see that #366 addressed zero sized stuff and the branch was merged. However, this apparently causes cmdstan to not anymore read in data serialized in this way as my example above demonstrates.

**Comment by maverickg** [comment]
I think cmdstan does not define how to deal with zero sized array/matrix/vector, so it's an issue of cmdstan.

**Comment by wds15** [comment]
I am not sure on that... the thing is that using rstan 2.12 one can export zero-sized arrays with stan_rdump which can be read with cmdstan 2.13. However, using rstan 2.13 creates files which will trigger cmdstan 2.13 to stop accepting that file.

So rstan 2.13 did change its behavior in how zero-sized arrays are exported. This change has not been accommodated for with cmdstan 2.13.

Should I open a ticket in the cmdstan repo?

**Comment by syclik** [comment]
Please open one in stan. It's in dump.hpp.

**Comment by wds15** [comment]
This one became obsolete as it got solved with 2.14. I think the change in the main stan repo fixed things.

**Comment by bob-carpenter** [comment]
Right --- Jiqiang fixed it in Stan and documented the new syntax.
